lib/value/src/kind/comparison.rs

```diff
@@ -0,0 +1,480 @@
+            return false;
+        }
+
+        let h = self.is_array();
+        if !(!h || !a && !b && !c && !d && !e && !f && !g) {
+            return false;
+        }
+
+        let i = self.is_object();
+        if !(!i || !a && !b && !c && !d && !e && !f && !g && !h) {
+            return false;
+        }
+
+        true
+    }
+
+    /// Check if `self` is a superset of `other`.
+    ///
+    /// Meaning, if `other` has a type defined as valid, then `self` needs to have it defined as
+    /// valid as well.
+    ///
+    /// Collection types are recursively checked (meaning, known fields in `self` also need to be
+    /// a superset of `other`.
+    #[must_use]
+    pub fn is_superset(&self, other: &Self) -> bool {
+        if let (None, Some(_)) = (self.bytes, other.bytes) {
+            return false;
+        };
+
+        if let (None, Some(_)) = (self.integer, other.integer) {
+            return false;
+        };
+
+        if let (None, Some(_)) = (self.float, other.float) {
+            return false;
+        };
+
+        if let (None, Some(_)) = (self.boolean, other.boolean) {
+            return false;
+        };
+
+        if let (None, Some(_)) = (self.timestamp, other.timestamp) {
+            return false;
+        };
+
+        if let (None, Some(_)) = (self.regex, other.regex) {
+            return false;
+        };
+
+        if let (None, Some(_)) = (self.null, other.null) {
+            return false;
+        };
+
+        match (self.array.as_ref(), other.array.as_ref()) {
+            (None, Some(_)) => return false,
+            (Some(lhs), Some(rhs)) if !lhs.is_superset(rhs) => return false,
+            _ => {}
+        };
+
+        match (self.object.as_ref(), other.object.as_ref()) {
+            (None, Some(_)) => return false,
+            (Some(lhs), Some(rhs)) if !lhs.is_superset(rhs) => return false,
+            _ => {}
+        };
+
+        true
+    }
+
+    /// Check if `self` intersects `other`.
+    ///
+    /// Returns `true` if there are type states common to both `self` and `other`.
+    #[must_use]
+    pub fn intersects(&self, other: &Self) -> bool {
+        if self.is_bytes() && other.is_bytes() {
+            return true;
+        }
+
+        if self.is_integer() && other.is_integer() {
+            return true;
+        }
+
+        if self.is_float() && other.is_float() {
+            return true;
+        }
+
+        if self.is_boolean() && other.is_boolean() {
+            return true;
+        }
+
+        if self.is_timestamp() && other.is_timestamp() {
+            return true;
+        }
+
+        if self.is_regex() && other.is_regex() {
+            return true;
+        }
+
+        if self.is_null() && other.is_null() {
+            return true;
+        }
+
+        if self.is_array() && other.is_array() {
+            return true;
+        }
+
+        if self.is_object() && other.is_object() {
+            return true;
+        }
```

Comment:
Where / how do we use the `intersects` method? I'd like to have some context to reason about the collection cases.

Reply:
In VRL, you can search the existing `type_def` in master, or wait for the follow-up PR.